common/src/main/java/io/netty/util/DefaultAttributeMap.java

```diff
@@ -17,143 +17,170 @@
-                }
-                curr = curr.next;
+    private <T> void removeAttributeIfMatch(AttributeKey<T> key, DefaultAttribute<T> value) {
+        for (;;) {
+            final DefaultAttribute[] attributes = this.attributes;
+            final int count = attributes.length;
+            final int index = searchAttributeByKey(attributes, 0, count, key);
+            if (index < 0) {
+                return;
+            }
+            final DefaultAttribute attribute = attributes[index];
+            assert attribute.key() == key;
+            if (attribute != value) {
+                return;
+            }
+            final int newCount = count - 1;
+            final DefaultAttribute[] newAttributes =
+                    newCount == 0? EMPTY_ATTRIBUTES : new DefaultAttribute[newCount];
+            // perform 2 bulk copies
+            System.arraycopy(attributes, 0, newAttributes, 0, index);
+            final int remaining = count - index - 1;
+            if (remaining > 0) {
+                System.arraycopy(attributes, index + 1, newAttributes, index, remaining);
+            }
+            if (ATTRIBUTES_UPDATER.compareAndSet(this, attributes, newAttributes)) {
+                return;
             }
-            return false;
         }
     }
 
-    private static int index(AttributeKey<?> key) {
-        return key.id() & MASK;
-    }
-
     @SuppressWarnings("serial")
     private static final class DefaultAttribute<T> extends AtomicReference<T> implements Attribute<T> {
 
+        private static final AtomicReferenceFieldUpdater<DefaultAttribute, DefaultAttributeMap> MAP_UPDATER =
+                AtomicReferenceFieldUpdater.newUpdater(DefaultAttribute.class,
+                                                       DefaultAttributeMap.class, "attributeMap");
         private static final long serialVersionUID = -2661411462200283011L;
 
+        private volatile DefaultAttributeMap attributeMap;
         // The head of the linked-list this attribute belongs to
-        private final DefaultAttribute<?> head;
         private final AttributeKey<T> key;
 
-        // Double-linked list to prev and next node to allow fast removal
-        private DefaultAttribute<?> prev;
-        private DefaultAttribute<?> next;
-
-        // Will be set to true one the attribute is removed via getAndRemove() or remove()
-        private volatile boolean removed;
-
-        DefaultAttribute(DefaultAttribute<?> head, AttributeKey<T> key) {
-            this.head = head;
+        DefaultAttribute(DefaultAttributeMap attributeMap, AttributeKey<T> key) {
+            this.attributeMap = attributeMap;
             this.key = key;
         }
 
-        // Special constructor for the head of the linked-list.
-        DefaultAttribute() {
-            head = this;
-            key = null;
-        }
-
         @Override
         public AttributeKey<T> key() {
             return key;
         }
 
+        public boolean removed() {
```

Comment:
remove `public` and rename to `isRemoved()` ?

Reply:
done!